cmd/relayer/main.go

```diff
@@ -0,0 +1,426 @@
+		},
+	}
+
+	errInvalidAddress       = errors.New("invalid forwarder address")
+	errNoEthereumPrivateKey = errors.New("must provide ethereum private key with --key")
+)
+
+func main() {
+	app := &cli.App{
+		Name:                 "relayer",
+		Usage:                "Ethereum transaction relayer",
+		Version:              cliutil.GetVersion(),
+		Flags:                flags,
+		Action:               run,
+		EnableBashCompletion: true,
+		Suggest:              true,
+	}
+
+	if err := app.Run(os.Args); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func setLogLevels(c *cli.Context) error {
+	const (
+		levelError = "error"
+		levelWarn  = "warn"
+		levelInfo  = "info"
+		levelDebug = "debug"
+	)
+
+	level := c.String(flagLog)
+	switch level {
+	case levelError, levelWarn, levelInfo, levelDebug:
+	default:
+		return fmt.Errorf("invalid log level %q", level)
+	}
+
+	_ = logging.SetLogLevel("main", level)
+	_ = logging.SetLogLevel("relayer", level)
+	_ = logging.SetLogLevel("rpc", level)
+	_ = logging.SetLogLevel("p2pnet", "debug")
+	return nil
+}
+
+func run(c *cli.Context) error {
+	err := setLogLevels(c)
+	if err != nil {
+		return err
+	}
+
+	port := uint16(c.Uint(flagRPCPort))
+	endpoint := c.String(flagEndpoint)
+	ec, err := ethclient.Dial(endpoint)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(c.Context)
+	defer cancel()
+
+	chainID, err := ec.ChainID(ctx)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("starting relayer with ethereum endpoint %s and chain ID %s", endpoint, chainID)
+
+	config, err := common.ConfigFromChainID(chainID)
+	if err != nil {
+		return err
+	}
+
+	key, err := getPrivateKey(c.String(flagKey))
+	if err != nil {
+		return err
+	}
+
+	// set the forwarder address from the config, if it exists on that network
+	// however, if the --forwarder-address flag is set, that address takes precedence
+	var contractAddr string
+	if (config.ContractAddress != ethcommon.Address{}) {
+		contractAddr = config.ForwarderContractAddress.String()
+	}
+	addrFromFlag := c.String(flagForwarderAddress)
+	if addrFromFlag != "" {
+		contractAddr = addrFromFlag
+	}
+
+	deploy := c.Bool(flagDeploy)
+	if deploy && (addrFromFlag != "") {
+		return fmt.Errorf("flags --%s and --%s are mutually exclusive", flagDeploy, flagForwarderAddress)
+	}
+	if !deploy && (contractAddr == "") {
+		return fmt.Errorf("either --%s or --%s is required", flagDeploy, flagForwarderAddress)
+	}
+
+	forwarder, forwarderAddr, err := deployOrGetForwarder(
+		ctx,
+		contractAddr,
+		ec,
+		key,
+		chainID,
+	)
+	if err != nil {
+		return err
+	}
+
+	relayerCommission, err := cliutil.ReadUnsignedDecimalFlag(c, flagRelayerCommission)
+	if err != nil {
+		return err
+	}
+
+	if relayerCommission.Cmp(apd.New(1, -1)) > 0 {
+		return errors.New("relayer commission is too high: must be less than 0.1 (10%)")
+	}
+
+	// TODO: do we need to restrict potential commission values? eg. 1%, 1.25%, 1.5%, etc
+	// or should we just require a fixed value for now?
```

Comment:
I still like the idea of a fixed-value commission, since the relayer's work is independent of the amount being transferred.  Just like miner fees only depend on the size of the transaction, not the amount of money being transferred.

Reply:
I agree, I think it might make more sense potentially, let's discuss it. I went with percentages to start off since that's what other relayer networks tended to use

Reply:
The worse scenario, that we'd like to avoid is:
* User wants to do large swap and doesn't like the percentage based commission.
* User does tiny swap with a relayer to get gas fees for a subsequent swap. The swap is so small that the relayer loses money on it.
* User now has gas fees to do a large swap without a relayer.